[AST] Add mechanism to notify module loading to other clients

medismailben opened the pull request:
This patch adds an optional callback function member to the AST Context.
The callback gets invoked when a new module (or overlay module) gets
loaded.

This can be used for instance by other clients, such as lldb, to perform
actions when a module gets loaded, like showing progress to the end-user.

rdar://94165195

Signed-off-by: Med Ismail Bennani <medismail.bennani@gmail.com>

adrian-prantl commented on lib/AST/ASTContext.cpp:
```diff
@@ -2212,6 +2214,8 @@ ASTContext::getModule(ImportPath::Module ModulePath) {
   for (auto &importer : getImpl().ModuleLoaders) {
     if (ModuleDecl *M = importer->loadModule(moduleID.Loc, ModulePath)) {
+      if (callback)
+        callback(llvm::formatv("Loading module {0}", M->getNameStr()).str());
```
It could be nice if the module loaders could customize the message to indicate that e.g., a module is being compiled from source.
Do we need to change the callback for subsequent invocations? If not, should maybe each module loader's constructor take an optional callback parameter instead of loadModule()?

medismailben replied:
> It could be nice if the module loaders could customize the message to indicate that e.g., a module is being compiled from source.

I agree, but I'm thinking the module loaders could just have static string with a getter that would be used by the callback function.

> Do we need to change the callback for subsequent invocations? If not, should maybe each module loader's constructor take an optional callback parameter instead of loadModule()?

I guess you meant `getModule` not `loadModule`, right? If it's so, when `getModule` with call the module loader `loadModule` implementation, since it's virtual pure, we would need to change each module loader implementation of `loadModule` to make sure it calls the callback. To me it's safer to invoke the callback at the `ASTContext` level.

medismailben commented on include/swift/AST/ASTContext.h:
```diff
@@ -1108,27 +1108,37 @@ class ASTContext final {
   /// be returned.
   ///
   /// \returns The requested module, or NULL if the module cannot be found.
```
> And should we call it something more descriptive with "progress" in the name?

I named it "callback" specifically so it can be used generically.

adrian-prantl replied:
If you want the name to be generic, then maybe a more descriptive name would be "postModuleLoadHook/Callback"?

JDevlieghere commented:
Did you consider making the callback a member of the ASTContext? Do the functions take different callbacks? I'm curious because that would require only adding a function + a member instead of changing 6 existing function signatures.

adrian-prantl commented on lib/AST/ASTContext.cpp:
```diff
@@ -2212,6 +2214,8 @@ ASTContext::getModule(ImportPath::Module ModulePath) {
   auto moduleID = ModulePath[0];
   for (auto &importer : getImpl().ModuleLoaders) {
     if (ModuleDecl *M = importer->loadModule(moduleID.Loc, ModulePath)) {
+      if (callback)
```
> Why is the callback coming after the load and not before?

I think I understand why now. This the outer loop in ASTContext, and you don't know ahead of time whether the load will succeed. But you could still move it outside of the loop.

adrian-prantl commented:
All in all I'm not sure I like this version of the patch. If (most) dependencies are loaded through ASTContext, then we should only need to modify loadModule() since that's called by all the other functions eventually.
My idea from yesterday for letting the module loaders customize the message doesn't actually work, because we can't know ahead of time whether the import is going to succeed. So the best thing we can do is printing the name of the module and the fact that we're importing it.

adrian-prantl commented:
It would be useful if we had a way of communicating the compilation of Clang module dependencies of Clang modules,  because that is very slow, but IIUC, not even ClangImporter know about this as it's handled by Clang? We could try building with `-Rmodule-build` and listening to the Diagnostics, but let's not make the perfect the enemy of the good. Even just printing the toplevel imports is better than nothing.

medismailben commented:
Addressed @JDevlieghere & @adrian-prantl feedbacks

medismailben commented:
@swift-ci test

adrian-prantl commented on lib/AST/ASTContext.cpp:
```diff
@@ -2234,12 +2243,17 @@ ModuleDecl *ASTContext::getOverlayModule(const FileUnit *FU) {
   }
 
+  if (PreModuleImportCallback) {
+    SmallString<256> path;
```
This function will likely be relatively high up in the call stack, so I wouldn't allocate such a large local variable here. Maybe reduce it to something like 16 or factor this into a helper function?

medismailben commented:
@swift-ci test

JDevlieghere commented on include/swift/AST/ASTContext.h:
```diff
@@ -1107,6 +1115,9 @@ class ASTContext final {
   /// If a module by this name has already been loaded, the existing module will
   /// be returned.
   ///
+  /// \param ModulePath The module's \c ImportPath which describes
+  /// the name of the module being loaded, possibly including submodules.
+
```
Outdated?

medismailben replied:
Nah, the ModulePath argument was already here originally, so I added some documentation for it.

JDevlieghere commented on lib/AST/ASTContext.cpp:
```diff
@@ -2210,6 +2217,8 @@ ASTContext::getModule(ImportPath::Module ModulePath) {
 
   auto moduleID = ModulePath[0];
+  if (PreModuleImportCallback)
+    PreModuleImportCallback(moduleID.Item.str(), false /*=IsOverlay*/);
```
Shouldn't this be `/*IsOverlay=*/ false`?

adrian-prantl commented on include/swift/AST/ASTContext.h:
```diff
@@ -489,6 +493,10 @@ class ASTContext final {
 
+  /// Set the module import callback function.
+  void setPreModuleImportCallback(
+      std::function<bool(llvm::StringRef, bool)> Callback);
```
After having seen the LLDB patch — why not make this a `= {}` argument in the constructor?

medismailben replied:
In my understanding, `std::function` creates an empty function by default, so no need to initialize it.

JDevlieghere replied:
I think Adrian is suggesting to make the Callback an optional argument in the constructor.

nkcsgexi commented on include/swift/AST/ASTContext.h:
```diff
@@ -374,6 +374,10 @@ class ASTContext final {
 
+  /// An optional generic callback function invoked prior to importing a module.
+  mutable std::function<bool(llvm::StringRef ModuleName, bool IsOverlay)>
+      PreModuleImportCallback;
```
Nitpick: wrap this with `Optional` to make its optionality explicit.

JDevlieghere replied:
Is that a common pattern in the swift repository? At least in LLVM (and LLDB) I've never seen that. As Ismail pointed out in another comment, `std::function` is empty by default so adding the optional means you have to check both the optional as well as the function before you can call it. That seems needlessly verbose to me.

JDevlieghere replied:
Maybe initializing this inline with `= {}` alleviates your concern?

nkcsgexi replied:
> Maybe initializing this inline with = {} alleviates your concern?

This works too. Thanks!

medismailben replied:
Perfect! I'll go with Adrian's suggestion

nkcsgexi reviewed (commented):
Reasonable change with a minor comment.

medismailben committed:
[AST] Add mechanism to notify module loading to other clients

This patch adds an optional callback function member to the AST Context.
The callback gets invoked when a new module (or overlay module) gets
loaded.

This can be used for instance by other clients, such as lldb, to perform
actions when a module gets loaded, like showing progress to the end-user.

rdar://94165195

Signed-off-by: Med Ismail Bennani <medismail.bennani@gmail.com>

medismailben commented:
@swift-ci test

medismailben commented:
Cherry-picked to main in #59466
